### PR TITLE
Reset trainer variable `should_stop` when `fit` is called

### LINKED ISSUE #18727: EarlyStopping not updating it's value after resuming training
### Bug description

Early stopping callback does not update its best value after resuming training from a checkpoint.

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
ckpt_path = "some_path/66_best_model.ckpt"
es = EarlyStopping(
		monitor  ="valid/step/loss",
		min_delta=0,
		patience =100,
		verbose  =True,
		mode     ="min",
        check_on_train_epoch_end=False,
	)

trainer = pl.Trainer(callbacks=[es])
trainer.fit(model, datamodule=datamodule, ckpt_path= ckpt_path)
```

### Error messages and logs

Hi,
I have an EarlyStopping and ModelCheckpoint on the same metrics in my callbacks. Also, I continue an existing training.

It seems like EarlyStopping does not update its best metrics if we load a checkpoint (retraining).

I got this output:
```
Epoch 149, global step 50700: 'epoch' reached 149.00000 (best 149.00000), saving model to '/dlocal/run/19197/_model_save/149_duration_model.ckpt' as top 100
Epoch 150, global step 51038: 'valid/step/loss' reached 0.00266 (best 0.00266), saving model to '/dlocal/run/19197/_model_save/150_best_model.ckpt' as top 1
Epoch 151, global step 51376: 'valid/step/loss' was not in top 1
Epoch 152, global step 51714: 'valid/step/loss' was not in top 1
Epoch 153, global step 52052: 'valid/step/loss' was not in top 1
Epoch 154, global step 52390: 'valid/step/loss' was not in top 1
Epoch 155, global step 52728: 'valid/step/loss' was not in top 1
Epoch 156, global step 53066: 'valid/step/loss' was not in top 1
Epoch 157, global step 53404: 'valid/step/loss' was not in top 1
Epoch 158, global step 53742: 'valid/step/loss' was not in top 1
Epoch 159, global step 54080: 'valid/step/loss' was not in top 1
Epoch 159, global step 54080: 'epoch' reached 159.00000 (best 159.00000), saving model to '/dlocal/run/19197/_model_save/159_duration_model.ckpt' as top 100
Generation Loop: 100%|██████████| 100/100 [05:37<00:00,  3.38s/it]it]2:26<03:13,  3.40s/it]
Generation Loop: 100%|██████████| 100/100 [05:38<00:00,  3.38s/it]
Generation Loop: 100%|██████████| 100/100 [05:38<00:00,  3.39s/it]
Generation Loop: 100%|██████████| 100/100 [05:39<00:00,  3.40s/it]
Epoch 160, global step 54418: 'valid/step/loss' was not in top 1
Epoch 161, global step 54756: 'valid/step/loss' was not in top 1
Epoch 162, global step 55094: 'valid/step/loss' was not in top 1
Epoch 163, global step 55432: 'valid/step/loss' was not in top 1
Epoch 164, global step 55770: 'valid/step/loss' was not in top 1
Epoch 165, global step 56108: 'valid/step/loss' was not in top 1
[rank: 0] Monitored metric valid/step/loss did not improve in the last 100 records. Best score: 0.003. Signaling Trainer to stop.
[rank: 3] Monitored metric valid/step/loss did not improve in the last 100 records. Best score: 0.003. Signaling Trainer to stop.
[rank: 1] Monitored metric valid/step/loss did not improve in the last 100 records. Best score: 0.003. Signaling Trainer to stop.
[rank: 2] Monitored metric valid/step/loss did not improve in the last 100 records. Best score: 0.003. Signaling Trainer to stop.
Epoch 166, global step 56446: 'valid/step/loss' was not in top 1
```

As we can see, in epoch 150 the model indeed improves.
But in epoch 165 the model stops (Early stopping with patience of 100), which should not be the case.
In fact, I resumed training from `66_best_model.ckpt`. This ckpt file is in a specific path outside my project for reasons.
It looks like the best metrics did not change from this ckpt.

### Environment

_No response_

### More info

_No response_

cc @carmocca @awaelchli

### PR DESCRIPTION
**Pull request by ryan597**
## What does this PR do?
Reset trainer variable `should_stop` when `fit` is called

If fit is called after early stopping has already stopped training, then the model will not continue training as the trainer flag `should_stop` is currently not reset when fit is called.
Change this to reset `should_stop` every time fit is called

Fixes #18727 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19177.org.readthedocs.build/en/19177/

### CODE REVIEW

**Comment by ryan597** [comment]
seems this is failing on a test that is designed to make sure the trainer stays as `should_stop=True` related to #15708 

```python
@pytest.mark.parametrize(("min_epochs", "min_steps", "val_count"), [(3, None, 3), (None, 3, 2)])
def test_should_stop_triggers_validation_once(min_epochs, min_steps, val_count, tmp_path):
    """Regression test for issue #15708.

    Test that the request for `should_stop=True` only triggers validation when Trainer is allowed to stop
    (min_epochs/steps is satisfied).

    """
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=tmp_path,
        num_sanity_val_steps=0,
        limit_val_batches=2,
        limit_train_batches=2,
        max_epochs=3,
        min_epochs=min_epochs,
        min_steps=min_steps,
        enable_model_summary=False,
        enable_checkpointing=False,
    )
    trainer.should_stop = True  # Request to stop before min_epochs/min_steps are reached
    trainer.fit_loop.epoch_loop.val_loop.run = Mock()
    trainer.fit(model)
    assert trainer.fit_loop.epoch_loop.val_loop.run.call_count == val_count
```

**Comment by ryan597** [comment]
I have changed the above test to use an `EarlyStopping` condition instead of setting the flag through `trainer.should_stop=True` such that this test now passes with the following 

```python
+    class NewBoring(BoringModel):
+        def training_step(self, batch, batch_idx):
+            self.log("loss", self.step(batch))
+            return {"loss": self.step(batch)}
+
-    model = BoringModel()
+    model = NewBoring()
+    # create a stopping condition with a high threshold so it triggers immediately
+    # check the condition before validation so the count is unaffected
+    stopping = EarlyStopping(monitor="loss", check_on_train_epoch_end=True, stopping_threshold=100)
     trainer = Trainer(
        default_root_dir=tmp_path,
        num_sanity_val_steps=0,
        limit_val_batches=2,
        limit_train_batches=2,
        max_epochs=3,
        min_epochs=min_epochs,
        min_steps=min_steps,
        enable_model_summary=False,
        enable_checkpointing=False,
        callbacks=[stopping],
    )
-   trainer.should_stop = True  # Request to stop before min_epochs/min_steps are reached
    trainer.fit_loop.epoch_loop.val_loop.run = Mock()
    trainer.fit(model)
    assert trainer.fit_loop.epoch_loop.val_loop.run.call_count == val_count

```

**Review by Borda — APPROVED** [review]

**Comment by qqueing** [comment]
is this PR in progress??
